### ISSUE TITLE
TypeVars for implicit lambdas can be tricked based on a single Callable

### ISSUE THREAD

**Issue by max-muoto** [microsoft/pyright#7927]
```python
from collections.abc import Callable
from typing import TypeVar

T = TypeVar("T", bool, float)

class MyClass:
    def map(self, func: Callable[[float], float] | Callable[[bytes], bytes]): ...

def foo(x: T) -> T: ...

MyClass().map(lambda x: foo(x))
```

Here, despite `T` being limited to either `bool | float`, we're possibly to pass in `bytes`. It seems only one of the unioned callables (regardless of the order) needs to match to trick Pyright into accepting this.

[Pyright Playground](https://pyright-play.net/?pyrightVersion=1.1.363&pythonVersion=3.12&strict=true&code=GYJw9gtgBAxmA28CmMAuBLMA7AzgOgEMAjGKdCABzBFSgGEDFjkAoUSKVATwvSwHMylarQAqPJADUCIFi1FQAvFHEUpMgBQAiUVoA0UImAQHg8MAVQBKOTHgEcOKAFkude44BcLKL6gATJGAoCAIKDRwkeGBTAFcsGE96RnsiZABtdLMLVABdU3NLXKgAH2SmNKRMoi5UJBx8w1r63KskvA65QODgYw0ADyTRKygAWgA%2BFXbOlld3BxwNKzxQ8PsIIn8CKEGoXrABqxs5E6A)

Note: I think this has been an issue for a while, not a regression with the improved union callable 

Version: 1.1.363

**Comment by max-muoto** [comment]
@erictraut Unrelated, and can create a separate issue for this, but Pylance uses the first callable for the type, even though I believe it's actually the union:

![CleanShot 2024-05-15 at 13 54 31](https://github.com/microsoft/pyright/assets/41130755/76565237-047c-4d31-a609-9b9bc7f805e0)

**Comment by erictraut** [comment]
I don't see any bug here. Pyright doesn't report any errors here, and I think that's correct. Mypy likewise doesn't report any errors.

Let's remove the TypeVar from the equation because I think it's needlessly complicating things. Instead, let's assume that foo is defined as simply: `def foo(x: float) -> float`. That would make `lambda x: foo(x)` compatible with `Callable[[float], float]` which makes it compatible with the `func` parameter type in the `MyClass.map` method.

Incidentally, the constrained TypeVar you've defined looks suspect to me. One of the constraints (`bool`) is a subtype of the other constraint (`float`). The typing spec is not clear on what should happen in this case. This issue is on a long list of issues that I'd like to have clarified in the typing spec. My preference is to make it illegal to specify overlapping constraints like this, but we'll need to see what the consensus of the community is. In any case, I recommend against defining constrained TypeVars with overlapping constraint types.

**Comment by max-muoto** [comment]
> I don't see any bug here. Pyright doesn't report any errors here, and I think that's correct. Mypy likewise doesn't report any errors.
> 
> Let's remove the TypeVar from the equation because I think it's needlessly complicating things. Instead, let's assume that foo is defined as simply: `def foo(x: float) -> float`. That would make `lambda x: foo(x)` compatible with `Callable[[float], float]` which makes it compatible with the `func` parameter type in the `MyClass.map` method.
> 
> Incidentally, the constrained TypeVar you've defined looks suspect to me. One of the constraints (`bool`) is a subtype of the other constraint (`float`). The typing spec is not clear on what should happen in this case. This issue is on a long list of issues that I'd like to have clarified in the typing spec. My preference is to make it illegal to specify overlapping constraints like this, but we'll need to see what the consensus of the community is. In any case, I recommend against defining constrained TypeVars with overlapping constraint types.

Thanks for the clarification! Without any subtypes, I assumed this should cause issues though, no?

```python
from collections.abc import Callable
from typing import TypeVar

from decimal import Decimal


class Container: ...


T = TypeVar("T", Container, Decimal)


class MyClass:
    def map(self, func: Callable[[str], str] | Callable[[Decimal], Decimal]): ...


def foo(x: T) -> T: ...


MyClass().map(lambda x: foo(x))

```

**Comment by erictraut** [comment]
No, why would that generate an error? The type of the `lambda` expression in this example is `(x: Decimal) -> Decimal` which is compatible with the annotated type of the `func` parameter. There's no type violation here.

**Comment by max-muoto** [comment]
> No, why would that generate an error? The type of the `lambda` expression in this example is `(x: Decimal) -> Decimal` which is compatible with the annotated type of the `func` parameter. There's no type violation here.

Sorry - as I modified a bit. I assumed the type of x in the lambda would be `str | Container` which would be incompatible with `Container`, it seems I am wrong in this assumption though.

**Comment by max-muoto** [comment]
For my personal understanding, is there a reason this example no longer works in 1.1.363 though? https://pyright-play.net/?pyrightVersion=1.1.363&pythonVersion=3.12&strict=true&code=GYJw9gtgBAxmA28CmMAuBLMA7AzgOgEMAjGKdCABzBFSgGEDFjkAoUSKVATwvSwHMylarQCCWLiykx4BHDigBZLnVnyAXCyjaoAEyTAoEAhQAUOJPGAAaKMACuWGOvqNZRZAG1POVCAC6tr4BUAA%2BrkweSN7iXIFQwf4AlC54aVIs%2BobAYGCmAB4ufKhJUAC0AHwJfqnpLMqqcjimSXjGZrIQRLoEUIV2uQVJSUA

**Comment by max-muoto** [comment]
I had a case like this, in my codebase, where previously the `Any` would get matched to the `int` in `foo`.

**Comment by erictraut** [comment]
That code didn't type check in previous versions of pyright either. You can use the version control in the pyright playground to confirm.

Evaluation of lambdas requires bidirectional type inference. In cases where the "expected type" consists of a union, pyright attempts to filter the subtypes of the union to remove any non-callalble types. If there are multiple callable types, it picks one of them to use for bidirectional type inference. To make this deterministic, it uses an internal sorting mechanism to sort the candidates. The sorting is somewhat arbitrary, since there is no well-defined order for types; the goal is to simply make it deterministic so union ordering doesn't change type checking behaviors. Perhaps you're recalling some other union of callables where the callable with the `Any` parameter happened to be sorted first. In any case, I wouldn't recommend creating a union of `Callable[[str], str]` and `Callable[[Any], str]`. Since the former is a subtype of the latter, you should simply use `Callable[[Any], str]`. Creating a union serves only to add complexity and confuse the type checker.

**Comment by max-muoto** [comment]
> That code didn't type check in previous versions of pyright either. You can use the version control in the pyright playground to confirm.
> 
> Evaluation of lambdas requires bidirectional type inference. In cases where the "expected type" consists of a union, pyright attempts to filter the subtypes of the union to remove any non-callalble types. If there are multiple callable types, it picks one of them to use for bidirectional type inference. To make this deterministic, it uses an internal sorting mechanism to sort the candidates. The sorting is somewhat arbitrary, since there is no well-defined order for types; the goal is to simply make it deterministic so union ordering doesn't change type checking behaviors. Perhaps you're recalling some other union of callables where the callable with the `Any` parameter happened to be sorted first. In any case, I wouldn't recommend creating a union of `Callable[[str], str]` and `Callable[[Any], str]`. Since the former is a subtype of the latter, you should simply use `Callable[[Any], str]`. Creating a union serves only to add complexity and confuse the type checker.

Did you try on 1.1.360, or is there something wrong here?

![CleanShot 2024-05-15 at 19 17 02](https://github.com/microsoft/pyright/assets/41130755/cabff16f-a124-40c2-b231-2b8bf40ed9a0)

**Comment by erictraut** [comment]
> Did you try on 1.1.360

No, I went back only two versions. It looks like something changed between 1.1.360 and 1.1.361. I don't consider it a bug though, for the reasons I explained above.

**Comment by max-muoto** [comment]
> > Did you try on 1.1.360
> 
> 
> 
> No, I went back only two versions. It looks like something changed between 1.1.360 and 1.1.361. I don't consider it a bug though, for the reasons I explained above.

-Sounds good! It's from a third party library so will just put up a PR tweaking it based on this, thanks.
